### PR TITLE
aws lambda enhanced monitoring

### PR DESCRIPTION
**Pull request by muzzamilinovaqo**

### CODE REVIEW

File: plugins/aws/lambda/lambdaEnhancedMonitoring.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            }
+
+            if (!listFunctions.data.length) {
+                helpers.addResult(results, 0, 'No Lambda functions found', region);
+                return rcb();
+            }
+
+            for (var lambdaFunc of listFunctions.data) {
+                if (!lambdaFunc.FunctionName) continue;
+                var resource = lambdaFunc.FunctionName;
```
Comment (fatima99s) [inline plugins/aws/lambda/lambdaEnhancedMonitoring.js:40]: resource should be the arn . store the function name in a seperate variable for the post call

File: plugins/aws/lambda/lambdaEnhancedMonitoring.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                var functionInfo = helpers.addSource(cache, source, ['lambda', 'getFunction', region, resource]);
+                
+                if (functionInfo && functionInfo.data && functionInfo.data.Configuration &&
```
Comment (fatima99s) [inline plugins/aws/lambda/lambdaEnhancedMonitoring.js:43]: check the checks for error and null data

File: plugins/aws/lambda/lambdaEnhancedMonitoring.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Lambda Enable Enhanced Monitoring',
+    category: 'Lambda',
+    domain: 'Serverless',
+    severity: 'Medium',
+    description: 'Ensure that AWS Lambda functions enhanced monitoring is enabled.',
+    more_info: 'Enhanced monitoring using Amazon CloudWatch Lambda Insights help you to monitor, troubleshoot, and optimize your functions.',
```
Comment (fatima99s) [inline plugins/aws/lambda/lambdaEnhancedMonitoring.js:10]: make it more descriptive

**Review by fatima99s — CHANGES_REQUESTED** [review]

File: plugins/aws/lambda/lambdaEnhancedMonitoring.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Lambda Enable Enhanced Monitoring',
```
Comment (fatima99s) [inline plugins/aws/lambda/lambdaEnhancedMonitoring.js:5]:
```suggestion
    title: 'Lambda Enhanced Monitoring Enabled',
```

File: plugins/aws/lambda/lambdaEnhancedMonitoring.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                helpers.addResult(results, 0, 'No Lambda functions found', region);
+                return rcb();
+            }
+
+            for (var lambdaFunc of listFunctions.data) {
+
+                if (!lambdaFunc.FunctionArn) continue;
+                var resource = lambdaFunc.FunctionArn;
+
+                var functionInfo = helpers.addSource(cache, source, ['lambda', 'getFunction', region, lambdaFunc.FunctionName]);
```
Comment (fatima99s) [inline plugins/aws/lambda/lambdaEnhancedMonitoring.js:43]:
```suggestion
                var functionInfo = helpers.addSource(cache, source, 
                ['lambda', 'getFunction', region, lambdaFunc.FunctionName]);
```

File: plugins/aws/lambda/lambdaEnhancedMonitoring.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            }
+
+            if (!listFunctions.data.length) {
+                helpers.addResult(results, 0, 'No Lambda functions found', region);
+                return rcb();
+            }
+
+            for (var lambdaFunc of listFunctions.data) {
+
+                if (!lambdaFunc.FunctionArn) continue;
```
Comment (fatima99s) [inline plugins/aws/lambda/lambdaEnhancedMonitoring.js:40]:
```suggestion
                if (!lambdaFunc.FunctionArn || !lambdaFunc.FunctionName) continue;
```

**Review by fatima99s — CHANGES_REQUESTED** [review]

File: plugins/aws/lambda/lambdaEnhancedMonitoring.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Lambda Enhanced Monitoring Enabled',
+    category: 'Lambda',
+    domain: 'Serverless',
+    severity: 'Medium',
+    description: 'Ensure that AWS Lambda functions enhanced monitoring is enabled.',
```
Comment (alphadev4) [inline plugins/aws/lambda/lambdaEnhancedMonitoring.js:9]:
```suggestion
    description: 'Ensure that Lambda functions have enhanced monitoring enabled.',
```

File: plugins/aws/lambda/lambdaEnhancedMonitoring.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+                if (!lambdaFunc.FunctionArn || !lambdaFunc.FunctionName) continue;
+                var resource = lambdaFunc.FunctionArn;
+
+                var functionInfo = helpers.addSource(cache, source, 
+                    ['lambda', 'getFunction', region, lambdaFunc.FunctionName]);
+
+                if (!functionInfo || functionInfo.err || !functionInfo.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to query for Lambda function Information: ${helpers.addError(functionInfo)}`, region,resource);
```
Comment (alphadev4) [inline plugins/aws/lambda/lambdaEnhancedMonitoring.js:48]:
```suggestion
                        `Unable to query for Lambda function details: ${helpers.addError(functionInfo)}`, region, resource);
```

File: plugins/aws/lambda/lambdaEnhancedMonitoring.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Lambda Enhanced Monitoring Enabled',
+    category: 'Lambda',
+    domain: 'Serverless',
+    severity: 'Medium',
+    description: 'Ensure that AWS Lambda functions enhanced monitoring is enabled.',
+    more_info: 'Amazon CloudWatch Lambda Insights is a monitoring and troubleshooting service for serverless applications running on Amazon Lambda. The service collects, aggregates, and summarizes system-level metrics including CPU, memory, disk, and network usage. It helps you to monitor, troubleshoot, and optimize your functions.',
+    link: 'https://docs.aws.amazon.com/lambda/latest/dg/monitoring-insights.html',
```
Comment (alphadev4) [inline plugins/aws/lambda/lambdaEnhancedMonitoring.js:11]:
```suggestion
    link: 'https://docs.aws.amazon.com/lambda/latest/dg/monitoring-insights.html#monitoring-insights-enabling-console',
```

File: plugins/aws/lambda/lambdaEnhancedMonitoring.spec.js [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+                    data: lambdaData
+                }
+            },
+            getFunction: functionInfoData
+
+        }
+    };
+};
+
+describe('Lambda Enable Enhanced Monitoring', function () {
```
Comment (alphadev4) [inline plugins/aws/lambda/lambdaEnhancedMonitoring.spec.js:19]:
```suggestion
describe('Lambda Enhanced Monitoring Enabled', function () {
```

File: plugins/aws/lambda/lambdaEnhancedMonitoring.spec.js [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+                {
+                    "FunctionName": "test-lambda",
+                    "FunctionArn": "arn:aws:lambda:us-east-1:000011112222:function:test-lambda"
+                }
+            ];
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(3);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Unable to query for Lambda function Information');
```
Comment (alphadev4) [inline plugins/aws/lambda/lambdaEnhancedMonitoring.spec.js:133]:
```suggestion
                expect(results[0].message).to.include('Unable to get Lambda function details:');
```

File: plugins/aws/lambda/lambdaEnhancedMonitoring.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+                        `Unable to get Lambda function details: ${helpers.addError(functionInfo)}`, region, resource);
+                    continue;
+                }
+                
+                if (functionInfo.data && 
+                    functionInfo.data.Configuration && 
+                    functionInfo.data.Configuration.Layers && 
+                    functionInfo.data.Configuration.Layers[0] &&
+                    functionInfo.data.Configuration.Layers[0].Arn) {
+                    helpers.addResult(results, 0, 'Lambda functions has enhanced monitoring enabled', region, resource);
```
Comment (alphadev4) [inline plugins/aws/lambda/lambdaEnhancedMonitoring.js:57]:
```suggestion
                    helpers.addResult(results, 0, 'Lambda function has enhanced monitoring enabled', region, resource);
```

File: plugins/aws/lambda/lambdaEnhancedMonitoring.spec.js [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+                        }
+                    }
+                }
+            };
+
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Lambda functions has enhanced monitoring enabled');
```
Comment (alphadev4) [inline plugins/aws/lambda/lambdaEnhancedMonitoring.spec.js:78]:
```suggestion
                expect(results[0].message).to.include('Lambda function has enhanced monitoring enabled');
```

**Review by alphadev4 — APPROVED** [review]

**Pull request merged**